### ISSUE TITLE
Namespaced policies

### ISSUE THREAD

**Issue by assembler** [varvet/pundit#12]
The app I'm working on has two sections: user and admin. There are multiple types of users and multiple types of admins, and different authentication solutions are used for each of them.

Although all the application models are used by both admins and users, policies are very different. Handling both users and admins and all distinct types within single policy was not so clear and full of `if`s. 

The way I handled this was to dynamically define `policy_class` on every `ActiveRecord` descendant in before filter. Something like this:

``` ruby
def set_policy_namespace
  m = self.class.name.split("::").first
  active_record_descendants.each do |ar|
    ar.class.instance_eval do
      define_method(:policy_class) do
        "::Policies::#{m}::#{self.name}Policy"
      end
    end
  end
end

def active_record_descendants
  Rails.application.eager_load! unless Rails.application.config.cache_classes
  ActiveRecord::Base.descendants
end
```

So, if I'm in `Admin::HomeController`, `Project` model will have `policy_class` redefined to return `::Policies::Admin::ProjectPolicy`. 

I've tried to demonstrate valid use case for this. Do you have any thoughts on how can this be made easier? I'm not sure whether or not it should be a part of gem (since it is somewhat rare).. I'm just asking for opinion.

thanks

**Comment by coryodaniel** [comment]
I just released a gem the other day. It is compatible with Pundit's policies and allows you to namespace based on controller, model or both. Check it out if you still need this functionality. [Regulator](https://github.com/coryodaniel/regulator)

**Comment by andreimoment** [comment]
Thank you!

Hope you can add "Use" section with a few examples to the readme... The
tests don't speak clearly to me.

**Comment by coryodaniel** [comment]
Will do this Sunday. No problem!

**Comment by andreimoment** [comment]
And -- I appreciate your releasing. It. Model-based permissions feel
strange for me as most of my apps have Admin-namespaced controllers. I have
an app ready to use this !

**Comment by martinverdejo** [comment]
how do I implement the `policy:` option from [here](https://github.com/elabs/pundit/issues/12#issuecomment-33827510)?

I have this working:
```raise Pundit::NotAuthorizedError unless Admin::HomePolicy.new(current_user, :home).index?```

but not this:
```authorize :home, :index?, policy: Admin::HomePolicy```

is this still the best approach?

**Comment by TangMonk** [comment]
Still do not know how to use, where is the documentation?

**Comment by thomasklemm** [comment]
@TangMonk Namespaced policies are not supported right now.

**Comment by TangMonk** [comment]
@thomasklemm, actually I figure it from older issues, please look at #382

**Comment by wayne5540** [comment]
## TL;DR: override `policy(record)` method

It took me a long time to figure out how to add namespace and do a controller based authorization, after tracing down the source code, it turns out very easy, I just need to override `policy(record)` from Pundit.

## My code:

I'll just post my solution as reference here, hope I can save someone's time.

In my base controller:

```rb
# app/controllers/api/v1/base_controller.rb
class Api::V1::BaseController < ApplicationController
  include Pundit
  after_action :verify_authorized

  private

  def policy(record)
    "#{controller_path.classify}Policy".constantize
  end
end
```

In my controller

```rb
# app/controllers/api/v1/auth/offers_controller.rb
class Api::V1::Auth::OffersController < Api::V1::Auth::BaseController
  def index
    authorize Offer
    # ...
  end
end
```

My policy

```rb
# app/policies/api/v1/auth/offer_policy.rb
class Api::V1::Auth::OfferPolicy < Api::V1::BasePolicy
  def index?
    true
  end
end
```

**Comment by cedv** [comment]
I came up with a different approach to the problem. Since Pundit does model-based authorization, no point fighting it.

Given a `Widget` model and the need for a `Admin` namespace policy, I created a new model `Admin::Widget` which just inherit `Widget`:

```ruby
# app/models/admin/widget.rb
class Admin::Widget < Widget
end
```

On the controller side, instead of using the `Widget` model for admin-related tasks, I use `Admin::Widget`:

```ruby
@widget = Admin::Widget.first
```

Policy-wise, I can have both `WidgetPolicy` and `Admin::WidgetPolicy` without any problem, keeping both authorization rules separate and neat.

**Comment by reentim** [comment]
Yet another option, in cases where defining `policy_class` on the model is undesirable:

```ruby
authorize OpenStruct.new(policy_class: ArbitraryPolicy, widget: @widget)
```

**Comment by coderliu** [comment]
Thanks @adamcrown  for his great work around.

Below is my modified base on his solution. [Link to gist](https://gist.github.com/coderliu/df69abc96cc0bc5098edb2ca4cab0651)
``` ruby
module ScopedPolicies
  extend ActiveSupport::Concern

  included do
    helper_method :authorize
    helper_method :policy_scope
  end

  # These are workarounds for the lack of support for namespacing in pundit
  # https://github.com/elabs/pundit/issues/12
  def controller_namespace
    @controller_namespace ||= self.class.to_s.sub(/::[A-z]*Controller/, '')
  end

  def authorize(record, query = nil)
    klass = "#{controller_namespace}::#{record.model_name}Policy".constantize
    policy = klass.new(current_user, record)
    query ||= "#{params[:action]}?"

    @_policy_authorized = true

    unless policy.public_send(query)
      error = Pundit::NotAuthorizedError.new("not allowed to #{query} this #{record}")
      error.query, error.record, error.policy = query, record, policy
      raise error
    end

    true
  end

  def policy_scope(scope)
    klass = "#{controller_namespace}::#{scope.model_name}Policy::Scope".constantize
    policy = klass.new(current_user, scope)

    @_policy_scoped = true

    policy.resolve
  end
end
```

**Comment by goulvench** [comment]
A new take for this old issue:
```ruby
# app/controllers/application_controller.rb
class ApplicationController < ActionController::Base

  protected

  def namespaced? = self.class.name.include?("::")
  def namespace = namespaced? ? self.class.module_parent.to_s.downcase.to_sym : nil

  def policy(record) = super([namespace, record].flatten.uniq)
  def policy_scope(scope) = super([namespace, scope])
  def authorize(record, query = nil) = super([namespace, record], query)
end
```

The `policy` method can be called directly (in views for instance) or through `policy_scope`, which already includes the namespace in this case, so it's necessary to `flatten.uniq` the parameters before passing them to `super`.

Feel free to place this in a concern, like I did, and only load it where necessary.
```ruby
# frozen_string_literal: true

module NamespacedPolicies
  extend ActiveSupport::Concern

  def policy(record) = super([namespace, record].flatten.uniq)
  def policy_scope(scope) = super([namespace, scope])
  def authorize(record, query = nil) = super([namespace, record], query)
end
```
